beacon_node/rest_api/src/validator.rs

```diff
@@ -0,0 +1,149 @@
+use super::{success_response, ApiResult};
+use crate::{helpers::*, ApiError, UrlQuery};
+use beacon_chain::{BeaconChain, BeaconChainTypes};
+use bls::PublicKey;
+use hyper::{Body, Request};
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use store::Store;
+use types::beacon_state::EthSpec;
+use types::{BeaconBlock, BeaconState, Epoch, RelativeEpoch, Shard, Slot};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ValidatorDuty {
+    /// The validator's BLS public key, uniquely identifying them. _48-bytes, hex encoded with 0x prefix, case insensitive._
+    #[serde(rename = "validator_pubkey")]
+    pub validator_pubkey: String,
+    /// The slot at which the validator must attest.
+    #[serde(rename = "attestation_slot")]
+    pub attestation_slot: Option<Slot>,
+    /// The shard in which the validator must attest.
+    #[serde(rename = "attestation_shard")]
+    pub attestation_shard: Option<Shard>,
+    /// The slot in which a validator must propose a block, or `null` if block production is not required.
+    #[serde(rename = "block_proposal_slot")]
+    pub block_proposal_slot: Option<Slot>,
+}
+
+impl ValidatorDuty {
+    pub fn new() -> ValidatorDuty {
+        ValidatorDuty {
+            validator_pubkey: "".to_string(),
+            attestation_slot: None,
+            attestation_shard: None,
+            block_proposal_slot: None,
+        }
+    }
+}
+
+/// HTTP Handler to retrieve a the duties for a set of validators during a particular epoch
+pub fn get_validator_duties<T: BeaconChainTypes + 'static>(req: Request<Body>) -> ApiResult {
+    // Get beacon state
+    let beacon_chain = req
+        .extensions()
+        .get::<Arc<BeaconChain<T>>>()
+        .ok_or_else(|| ApiError::ServerError("Beacon chain extension missing".to_string()))?;
+    let _ = beacon_chain
+        .ensure_state_caches_are_built()
+        .map_err(|e| ApiError::ServerError(format!("Unable to build state caches: {:?}", e)))?;
+    let head_state = beacon_chain
+        .speculative_state()
+        .expect("This is legacy code and should be removed.");
+
+    // Parse and check query parameters
+    let query = UrlQuery::from_request(&req)?;
+
+    let current_epoch = head_state.current_epoch();
+    let epoch = match query.first_of(&["epoch"]) {
+        Ok((_, v)) => Epoch::new(v.parse::<u64>().map_err(|e| {
+            ApiError::InvalidQueryParams(format!("Invalid epoch parameter, must be a u64. {:?}", e))
+        })?),
+        Err(_) => {
+            // epoch not supplied, use the current epoch
+            current_epoch
+        }
+    };
+    let relative_epoch = RelativeEpoch::from_epoch(current_epoch, epoch)
+        .map_err(|e| ApiError::InvalidQueryParams(format!("Cannot get RelativeEpoch: {:?}", e)))?;
```

Comment:
We'll get this error if `epoch` is more than `1` epoch from `head_state`.

We could perhaps return an error message like "Epoch out of range"

Reply:
I have added clarification to this, but the `EpochTooHigh` error will be returned in that case ,which will be included in the error message.